ssl/ssl_sess.c

```diff
@@ -139,8 +139,11 @@ SSL_SESSION *ssl_session_dup(SSL_SESSION *src, int ticket)
     dest->references = 1;
 
     dest->lock = CRYPTO_THREAD_lock_new();
-    if (dest->lock == NULL)
+    if (dest->lock == NULL) {
+        OPENSSL_free(dest);
+        dest = NULL;
         goto err;
+    }
 
     if (!CRYPTO_new_ex_data(CRYPTO_EX_INDEX_SSL_SESSION, dest, &dest->ex_data))
         goto err;
```

Comment:
It seems like there is also an issue here, however. If this fails we will call X509_free on ss->peer, for example. Since the pointer is just copied from `src` the above `memcpy` call after this call fails, this will lead to a double free after the original session is eventually freed.